src/rust/engine/ui/src/display.rs

```diff
@@ -0,0 +1,307 @@
+  logs: VecDeque<String>,
+  running: bool,
+  is_tty: bool,
+  cursor_start: (u16, u16),
+  terminal_size: (u16, u16),
+}
+
+// TODO: Prescribe a threading/polling strategy for callers - or implement a built-in one.
+// TODO: Better error handling for .flush() and .write() failure modes.
+impl EngineDisplay {
+  pub fn for_stdout(indent_level: u16) -> EngineDisplay {
+    let write_handle = stdout();
+    let is_tty = termion::is_tty(&write_handle);
+
+    EngineDisplay {
+      sigil: '⚡',
+      divider: "▵".to_string(),
+      poll_interval_ms: Duration::from_millis(55),
+      padding: (0..indent_level)
+        .map({ |_| " " })
+        .collect::<Vec<&str>>()
+        .concat(),
+      terminal: if !is_tty {
+        Console::Pipe(write_handle)
+      } else {
+        match write_handle.into_raw_mode() {
+          Ok(t) => Console::Terminal(t),
+          Err(_) => Console::Pipe(stdout()),
+        }
+      },
+      action_map: HashMap::new(),
+      logs: VecDeque::new(),
+      running: false,
+      is_tty: is_tty,
+      // N.B. This will cause the screen to clear - but with some improved position
+      // tracking logic we could avoid screen clearing in favor of using the value
+      // of `EngineDisplay::get_cursor_pos()` as initialization here. From there, the
+      // trick would be to push the on-screen text screen up (ideally, prior to entering
+      // raw mode) via newline printing - and then ensuring the rendered text never
+      // overflows past the screen row bounds. Before we could safely do that tho, we
+      // probably want to plumb a signal handler for SIGWINCH so that ^L works to redraw
+      // the screen in the case of any display issues. Otherwise, simply clearing the screen
+      // as we've done here is the safest way to avoid terminal oddness.
+      cursor_start: (1, 1),
+      terminal_size: EngineDisplay::get_size(),
+    }
+  }
+
+  // Gets the current terminal's cursor position, if applicable.
+  fn get_cursor_pos(&mut self) -> (u16, u16) {
+    match self.terminal {
+      Console::Terminal(ref mut t) => match &t.cursor_pos() {
+        Ok((x, y)) => (*x, *y),
+        // N.B. Real TTY coordinates start at 1.
+        Err(_) => (0, 0),
+      },
+      Console::Pipe(_) => (0, 0),
+    }
+  }
+
+  // Gets the current terminal's width and height, if applicable.
+  fn get_size() -> (u16, u16) {
+    match termion::terminal_size() {
+      Ok((x, y)) => (x, y),
+      Err(_) => (0, 0),
+    }
+  }
+
+  // Whether or not the EngineDisplay is running (whether .start() has been called).
+  pub fn is_running(&self) -> bool {
+    self.running
+  }
+
+  // Sets the terminal size per-render for signal-free resize detection.
+  fn set_size(&mut self) {
+    self.terminal_size = EngineDisplay::get_size();
+  }
+
+  // Sets the terminal size for signal-free resize detection.
+  fn get_max_log_rows(&self) -> usize {
+    self.terminal_size.1 as usize - self.action_map.len() - 1
```

Comment:
yeah.. hopefully this case is rare by nature, but I suspect we'd want to fall back to non-tty mode and just scroll a summary as if a tty weren't there (since a tty with few lines is effectively useless anyway). added a TODO.